### PR TITLE
refactor: more flexible extrinsic unwrapping for ctype fetching

### PR DESCRIPTION
**Pull request by rflechtner**
## fixes https://github.com/KILTprotocol/ticket/issues/3119

While I hot fixed the current behaviour in https://github.com/KILTprotocol/sdk-js/pull/836 to allow extrinsics submitted via dispatchAs, I think it's a better strategy for the future to build on top of on an unwrapper function that returns nested calls contained within other extrinsics that can easily be amended. E.g., we could easily add unwrapping of proxy calls to the logic here and everything still works.

## How to test:

Ideally, fetch a CType with it.

## Checklist:

- [x] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)
    * Either PR or Ticket to update [the Docs](https://github.com/KILTprotocol/docs)
    * Link the PR/Ticket here

### CODE REVIEW

**Review by ntn-x2 — COMMENTED** [review]
I like your solution, very elegant. I am wondering if it can be structured a bit better with regards to functions responsibilities. Other than that, I like it a lot!

File: packages/credentials/src/ctype/CType.chain.ts [hunk 1]
```diff
@@ -168,71 +143,43 @@ export async function fetchFromChain(
...
+  const ctypeCreationCalls = extrinsicCalls.filter(
+    (c): c is GenericCall<typeof api.tx.ctype.add.args> =>
+      api.tx.ctype.add.is(c)
+  )
 
-  // If more than a call is present, it always considers the last one as the valid one.
-  const lastRightCTypeCreationCall = ctypeCallContent
+  // Re-create the ctype for each call identified to find the right ctype.
+  // If more than one matching call is present, it always considers the last one as the valid one.
+  const lastRightCTypeCreationCall = ctypeCreationCalls
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/credentials/src/ctype/CType.chain.ts:168]: we could, yes, which would then also help avoid that an id is computed twice

File: packages/asset-credentials/src/credentials/PublicCredential.chain.ts [hunk 2]
```diff
@@ -183,58 +164,40 @@ export async function fetchCredentialFromChain(
...
-    })
+  // only consider public_credentials::add calls
+  const publicCredentialCalls = extrinsicCalls.filter(
+    (c): c is GenericCall<typeof api.tx.publicCredentials.add.args> =>
+      api.tx.publicCredentials.add.is(c)
+  )
 
-  if (!lastRightCredentialCreationCall) {
+  // Re-create the issued public credential for each call identified to find the credential with the id we're looking for
+  const credentialInput = publicCredentialCalls.reduceRight<
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/asset-credentials/src/credentials/PublicCredential.chain.ts:177]: what would that look like and what would its responsibility be? Seems very specific to this one use case in this one function

File: packages/chain-helpers/src/blockchain/Blockchain.ts [hunk 3]
```diff
@@ -231,10 +233,17 @@ export function isBatch(
  *
  * @returns A list of {@link Call} nested according to the rules above.
  */
-export function flattenCalls(call: Call, api?: ApiPromise): Call[] {
-  if (isBatch(call, api)) {
+export function flattenCalls(call: IMethod, api?: ApiPromise): IMethod[] {
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:236]: Yes, a rename of the function + a complete list of functions that contain nested calls would be a better design for sure.

File: packages/asset-credentials/src/credentials/PublicCredential.chain.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply packages/asset-credentials/src/credentials/PublicCredential.chain.ts:177]: This becomes irrelevant if this is replaced by the logic used for public credentials anyway.

File: packages/chain-helpers/src/blockchain/Blockchain.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:236]: The name still fits the function as far as I'm concerned, do you want to rename it to mark the different behaviour? It is basically newly introduced in the v1 release though, given that it lived in a completely different package before

[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:236]: what would you suggest we call it?

File: packages/credentials/src/ctype/CType.chain.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/credentials/src/ctype/CType.chain.ts]: a6f9852ef6efbd9ae30be20fbb09f5ec51bf3007

File: packages/asset-credentials/src/credentials/PublicCredential.chain.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/asset-credentials/src/credentials/PublicCredential.chain.ts:177]: Does this address your concern then? a6f9852ef6efbd9ae30be20fbb09f5ec51bf3007

File: packages/chain-helpers/src/blockchain/Blockchain.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:236]: added proxy support in 405960fc8d7c9d1c1f1e173c69b035cea4098cda

**Review by ntn-x2 — APPROVED** [review]
Left one more minor comment, otherwise it looks good.

File: packages/chain-helpers/src/blockchain/Blockchain.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:236]: that's a good idea!

[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/chain-helpers/src/blockchain/Blockchain.ts:242]: 4b30fd40d8618bdd9f8ee4bc4a5ad7bcd2ba317f

**Pull request merged**